src/Microsoft.Android.Build.BaseTasks/AsyncTask.cs

```diff
@@ -0,0 +1,461 @@
+// https://github.com/xamarin/Xamarin.Build.AsyncTask/blob/db4ce14dacfef47435c238b1b681c124e60ea1a0/Xamarin.Build.AsyncTask/AsyncTask.cs
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using Microsoft.Build.Utilities;
+using Microsoft.Build.Framework;
+using System.Threading;
+using static System.Threading.Tasks.TaskExtensions;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.Android.Build.Tasks
+{
+	/// <summary>
+	/// Base class for tasks that need long-running cancellable asynchronous tasks 
+	/// that don't block the UI thread in the IDE.
+	/// </summary>
+	public abstract class AsyncTask : Task, ICancelableTask
+	{
+		readonly CancellationTokenSource cts = new CancellationTokenSource ();
+		readonly Queue logMessageQueue = new Queue ();
+		readonly Queue warningMessageQueue = new Queue ();
+		readonly Queue errorMessageQueue = new Queue ();
+		readonly Queue customMessageQueue = new Queue ();
+		readonly Queue telemetryMessageQueue = new Queue ();
+		readonly ManualResetEvent logDataAvailable = new ManualResetEvent (false);
+		readonly ManualResetEvent errorDataAvailable = new ManualResetEvent (false);
+		readonly ManualResetEvent warningDataAvailable = new ManualResetEvent (false);
+		readonly ManualResetEvent customDataAvailable = new ManualResetEvent (false);
+		readonly ManualResetEvent telemetryDataAvailable = new ManualResetEvent (false);
+		readonly ManualResetEvent taskCancelled = new ManualResetEvent (false);
+		readonly ManualResetEvent completed = new ManualResetEvent (false);
+		bool isRunning = true;
+		object eventlock = new object ();
+		int uiThreadId = 0;
+
+		/// <summary>
+		/// Indicates if the task will yield the node during tool execution.
+		/// </summary>
+		public bool YieldDuringToolExecution { get; set; }
+
+		/// <summary>
+		/// The cancellation token to notify the cancellation requests
+		/// </summary>
+		public CancellationToken CancellationToken => cts.Token;
+
+		/// <summary>
+		/// Gets the current working directory for the task, which is captured at task 
+		/// creation time from <see cref="Directory.GetCurrentDirectory"/>.
+		/// </summary>
+		protected string WorkingDirectory { get; private set; }
+
+		[Obsolete ("Do not use the Log.LogXXXX from within your Async task as it will Lock the Visual Studio UI. Use the this.LogXXXX methods instead.")]
+		private new TaskLoggingHelper Log => base.Log;
+
+		/// <summary>
+		/// Initializes the task.
+		/// </summary>
+		protected AsyncTask ()
+		{
+			YieldDuringToolExecution = false;
+			WorkingDirectory = Directory.GetCurrentDirectory ();
+			uiThreadId = Thread.CurrentThread.ManagedThreadId;
+		}
+
+		public void Cancel ()
+			=> taskCancelled.Set ();
+
+		protected void Complete (System.Threading.Tasks.Task task)
+		{
+			if (task.Exception != null) {
+				var ex = task.Exception.GetBaseException ();
+				LogError (ex.Message + Environment.NewLine + ex.StackTrace);
```

Comment:
@dellis1972: any recollection of why we logged the error this way instead of just logging `ex.ToString()`?

@dellis1972, @pjcollins: with the realization that this next comment goes well into "the perfect is the enemy of the good"… Should we move `AndroidAsyncTask.TaskPrefix` into this type so that we can use `LogCodedError()` keyed off of `TaskPrefix` here and elsewhere?

Reply:
> @dellis1972, @pjcollins: with the realization that this next comment goes well into "the perfect is the enemy of the good"… Should we move AndroidAsyncTask.TaskPrefix into this type so that we can use LogCodedError() keyed off of TaskPrefix here and elsewhere?

I think this goes back to the question above -- it probably does make most sense to merge the two classes and deal with the renaming/breakage that comes with that.

Reply:
We can merge them if we want, I was just trying to make the bumps a bit easier. But if its better in the long run then go for it 👍

Reply:
Merged the two, and updated this Exception logging function to use LogUnhandledException which will output an error code using TaskPrefix.